File: src/xSTIR/pSTIR/tests/test1.py
''' pSTIR tests
'''

## CCP PETMR Synergistic Image Reconstruction Framework (SIRF)
## Copyright 2015 - 2017 Rutherford Appleton Laboratory STFC
##
## This is software developed for the Collaborative Computational
## Project in Positron Emission Tomography and Magnetic Resonance imaging
## (http://www.ccppetmr.ac.uk/).
##
## Licensed under the Apache License, Version 2.0 (the "License");
##   you may not use this file except in compliance with the License.
##   You may obtain a copy of the License at
##       http://www.apache.org/licenses/LICENSE-2.0
##   Unless required by applicable law or agreed to in writing, software
##   distributed under the License is distributed on an "AS IS" BASIS,
##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##   See the License for the specific language governing permissions and
##   limitations under the License.

import math

from pSTIR import *

def test_failed(ntest, expected, actual, abstol, reltol):
    if abs(expected - actual) < abstol + reltol*expected:
        print('+++ test %d passed' % ntest)
        return 0
    else:
        print('+++ test %d failed' % ntest)
        return 1

def norm(v):
    vv = v*v
    nv = v.size
#    return vv.sum()/nv
    return math.sqrt(vv.sum()/nv)

# a function to compute the variance after conversion to double to avoid
# rounding problems with older numpy versions
def var(v):
    return v.astype(numpy.float64).var()

def main():

    failed = 0
    eps = 1e-4

    # create matrix to be used by the acquisition model
    matrix = RayTracingMatrix()
    matrix.set_num_tangential_LORs(2)

    # create acquisition model
    am = AcquisitionModelUsingMatrix()
    am.set_matrix(matrix)

    # locate the input data file folder
    data_path = petmr_data_path('pet')

    # PET acquisition data to be read from this file
    raw_data_file = existing_filepath(data_path, 'Utahscat600k_ca_seg4.hs')
    ad = AcquisitionData(raw_data_file)
    adata = ad.as_array()
    s = norm(adata)
    v = var(adata)
    failed += test_failed(1, 2.510818, s, 0, eps)
    failed += test_failed(2, 5.444323, v, 0, eps)
    #print('acquisitions mean sum of squares: %f, variance: %f' % (s, v))

    # create filter
    filter = TruncateToCylinderProcessor()

    # create initial image estimate
    image_size = (111, 111, 31)
    voxel_size = (3, 3, 3.375) # voxel sizes are in mm
    image = ImageData()
    image.initialise(image_size, voxel_size)
    image.fill(1.0)
    
    filter.apply(image)
    image_arr = image.as_array()
    s = norm(image_arr)
    v = var(image_arr)
    failed += test_failed(3, 0.876471, s, 0, eps)
    failed += test_failed(4, 0.178068, v, 0, eps)
    #print('image mean sum of squares: %f, variance: %f' % (s, v))

    # create prior
    prior = QuadraticPrior()
    prior.set_penalisation_factor(0.5)

    # set number of subsets
    num_subsets = 12

    # create objective function
    obj_fun = make_Poisson_loglikelihood(ad)
    obj_fun.set_acquisition_model(am)
    obj_fun.set_num_subsets(num_subsets)
    obj_fun.set_up(image)

    # select subset
    subset = 0

    # get sensitivity as ImageData
    ss_img = obj_fun.get_subset_sensitivity(subset)

    # get back projection of the ratio of measured to estimated acquisition data
    grad_img = obj_fun.get_backprojection_of_acquisition_ratio(image, subset)

    # get gradient of prior as ImageData
    pgrad_img = prior.get_gradient(image)

    # copy to Python arrays
    image_arr = image.as_array()
    ss_arr = ss_img.as_array()
    grad_arr = grad_img.as_array()
    pgrad_arr = pgrad_img.as_array()

    # update image data
    ss_arr[ss_arr < 1e-6] = 1e-6 # avoid division by zero
    update = grad_arr/(ss_arr + pgrad_arr/num_subsets)
    image_arr = image_arr*update

    s = norm(image_arr)
    v = var(image_arr)
    failed += test_failed(5, 0.012314, s, 0, eps)
    failed += test_failed(6, 0.000052, v, eps, eps)
    #print('image mean sum of squares: %f, variance: %f' % (s, v))
    s = norm(update)
    v = var(update)
    failed += test_failed(7, 3.846513, s, 0, eps)
    failed += test_failed(8, 14.775219, v, 0, eps)
    #print('update mean sum of squares: %f, variance: %f' % (s, v))
    s = norm(ss_arr)
    v = var(ss_arr)
    failed += test_failed(9, 27.990159, s, 0, eps)
    failed += test_failed(10, 207.401144, v, 0, eps)
    #print('sensitivity mean sum of squares: %f, variance: %f' % (s, v))
    s = norm(grad_arr)
    v = var(grad_arr)
    failed += test_failed(11, 98.049032, s, 0, eps)
    failed += test_failed(12, 9599.796540, v, 0, eps)
    #print('gradient mean sum of squares: %f, variance: %f' % (s, v))
    s = norm(pgrad_arr)
    v = var(pgrad_arr)
    failed += test_failed(13, 0.710633, s, 0, eps)
    failed += test_failed(14, 0.505000, v, 0, eps)
    #print('prior gradient mean sum of squares: %f, variance: %f' % (s, v))
    return failed

try:
    failed = main()
    if failed == 0:
        print('all tests passed')
        sys.exit(0)
    else:
        print('%d tests failed' % failed)
        sys.exit(failed)

except error as err:
    # display error information
    print('??? %s' % err.value)
    sys.exit(-1)

File: src/xSTIR/mSTIR/tests/test1.m
function [failed, ntests] = test1(record, engine)
% PET test set 1.

% SyneRBI Synergistic Image Reconstruction Framework (SIRF).
% Copyright 2015 - 2019 Rutherford Appleton Laboratory STFC.
% Copyright 2015 - 2020 University College London.
% 
% This is software developed for the Collaborative Computational
% Project in Synergistic Reconstruction for Biomedical Imaging (formerly CCP PETMR)
% (http://www.ccpsynerbi.ac.uk/).
% 
% Licensed under the Apache License, Version 2.0 (the "License");
% you may not use this file except in compliance with the License.
% You may obtain a copy of the License at
% http://www.apache.org/licenses/LICENSE-2.0
% Unless required by applicable law or agreed to in writing, software
% distributed under the License is distributed on an "AS IS" BASIS,
% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
% See the License for the specific language governing permissions and
% limitations under the License.

% Select and import SIRF MATLAB MR package so that SIRF MR objects can be 
% created in this function without using the prefix 'MR.'
if nargin < 2
    engine = [];
end
if nargin < 1
    record = false;
end
% import_str = set_up_PET(engine);
% eval(import_str)
PET = set_up_PET(engine);

test = sirf.Utilities.mTest('test1.txt', record);

% define raw data source
filename = 'Utahscat600k_ca_seg4.hs';
pathname = sirf.Utilities.examples_data_path('PET');
acq_data = PET.AcquisitionData(fullfile(pathname, filename));
s = acq_data.norm();
v = variance(acq_data);
test.check(s)
test.check(v)
disp("Printing AcqData info")
disp(acq_data.get_info())

image = PET.ImageData();
image_size = [111, 111, 31];
voxel_size = [3, 3, 3.375];
image.initialise(image_size, voxel_size)
image.fill(1.0)
filter = PET.TruncateToCylinderProcessor();
filter.apply(image)
s = image.norm();
v = variance(image);
test.check(s)
test.check(v)

prior = PET.QuadraticPrior();
prior.set_penalisation_factor(0.5)
prior.set_up(image);

matrix = PET.RayTracingMatrix();
matrix.set_num_tangential_LORs(2)

am = PET.AcquisitionModelUsingMatrix();
am.set_matrix(matrix)

num_subsets = 12;

obj_fun = PET.make_Poisson_loglikelihood(acq_data);
obj_fun.set_acquisition_model(am)
obj_fun.set_num_subsets(num_subsets)
obj_fun.set_up(image)

subset = 0;

sens_image = obj_fun.get_subset_sensitivity(subset);

grad_image = obj_fun.get_backprojection_of_acquisition_ratio...
             (image, subset);

prior_grad_image = prior.get_gradient(image);

image_array = image.as_array();
sens_array = sens_image.as_array();
grad_array = grad_image.as_array();
prior_grad_array = prior_grad_image.as_array();

denom = sens_array + prior_grad_array./num_subsets;
delta = 1e-6*max(abs(denom(:)));
denom(denom < delta) = delta; % avoid division by zero
update = grad_array./denom;
image_array = image_array.*update;

eps = 1e-4;
s = norm(update(:));
v = var(update(:));
delta = max(update(:))*eps;
test.check(s)
test.check(v, delta)

s = norm(image_array(:));
v = var(image_array(:));
delta = max(image_array(:))*eps;
test.check(s)
test.check(v, delta)

% Check that the storage scheme for a pre-initialised object
% doesn't change just because the default has changed.
temp = PET.AcquisitionData();
temp.set_storage_scheme('memory');
ad = PET.AcquisitionData(fullfile(pathname, filename));
temp.set_storage_scheme('file');
test.check_if_equal('memory', ad.get_storage_scheme());

failed = test.failed;
ntests = test.ntest;

if record
    fprintf('%d measurements recorded\n', ntests)
elseif failed == 0
    fprintf('all %d tests passed\n', ntests)
else
    fprintf('%d out of %d tests failed\n', failed, ntests)
end
end

function v = variance(x)
a = double(x.as_array());
v = var(a(:));
end
